### ISSUE TITLE
Deadlocks using TransactionalIndexedCollection

### ISSUE THREAD

**Issue by demwis** [npgall/cqengine#235]
I have the following initial set derived from the real one (CQEngine v3.3.0):
```
class Wrapper {    
  private static final Attribute<Item, String> FIELD1_ATTR = QueryFactory.nulllableAttribute("FIELD1", Item::getField1);
  private final IndexedCollection<Item> items = new TransactionalIndexedCollection<>(Item.class);
  
  public Wrapper() {
    items.addIndex(HashIndex.onAttribute(FIELD1_ATTR));
  }

  public int getCountByField1(String value) {
    try (ResultSet<Item> results = items.retrieve(QueryFactory.equal(FIELD1_ATTR, value)) {
      return results.size();
    }
  }

  public removeItem(Item item) {
    items.remove(item);
  }
}
```

The general problem is that when I pass null value to getCountByField1 method, it fails with NPE which leads to the unreleased lock:

> java.lang.NullPointerException
>   at com.g...simpleEqual.calcHashCode(Equal.java:78)
> ....
> at com.g...TransactionalIndexedCollection.retrieve(TransactionalIndexedCollection.java:373)
> ...
There is a following code in that place:
```
final Version thisVersion = acquireReadLockForCurrentVersion(); // 363 Inside of this method it locks on readLock
...
ResultSet<O> results = super.retrieve(query, queryOptions); // 373
```
So after the NPE was thrown there is no place where readLock is released.
After that when I try to remove the item using removeItem method it just hangs in WAITING (parking) state while is incrementing version:
> ...
> at com.g...TransactionalIndexedCollection.incrementVersion(TransactionalIndexedCollection.java:161)
> ...
On that line it tries to seize writeLock on currentVersion:
```
Version previousVersion = this.currentVersion;
this.currentVersion = new Version(...);
previousVersion.lock.writeLock().lock(); // 161
```
So I got deadlock here.

The second problem or better say question is why I can't pass null search parameter for nullable attribute of the collection. If index can store null value, why it doesn't allow to search by it?

Thank you for your attention. Looking forward to hearing from you soon.

**Comment by npgall** [comment]
Thanks for reporting this. I think I should elaborate in the documentation some more about why null handling is the way it is. I am also planning to address #223 which should prevent this issue you saw from happening. 

To check if a value is null, you need to use a `not(has())` query. That is the equivalent in CQEngine of the SQL query "IS NOT NULL".

CQEngine queries are similar to SQL queries in how null is handled, such that null values are not permitted in comparison operations. I can't say what the history/rationale for this decision was with SQL, but I can say that for CQEngine adopting this approach eliminated some ambiguity in how queries are evaluated and ordered. For example it is not clear how a query `lessThan(Car.PRICE, null))` should be evaluated. How should null be compared with non-null values, and then the question arises should objects with null values be returned at all. It is just simpler to not permit nulls in queries by default at all. 

So CQEngine and SQL both take the approach that null is not permitted in most queries, except for the special queries which are specifically for dealing with nulls: `has()` and `not(has())` (for CQEngine) and "IS NOT NULL" and "IS NULL" (for SQL).

So anyway, back to the deadlock issue you encountered. If you use a `not(has())` query instead, it should prevent this problem. However I totally agree the user experience here could definitely be improved. So I will see if the code can be improved here in case null is supplied, and also with #223 I am planning to throw a more informative exception message.

**Comment by demwis** [comment]
Thank you for your detailed reply, Niall. Me and my team hope TransactionalIndexedCollection will be fixed soon. For now we decided to switch to ConcurrentIndexedCollection.

**Comment by npgall** [comment]
I have just committed a fix for this in the commit above. 

Also, I have fixed a related issue #223 which will cause the constructors of query objects to now throw exceptions proactively, if null values are supplied for queries which do not support nulls. 

Going forward, those exception messages should now be more informative, recommending users to use `has()` query instead instead of ending up with bad behavior, or otherwise confusing behavior like this.

Thanks again for reporting this, and especially for deep diving into the code to pinpoint the issue. The issue was exactly where you said. Very helpful, cheers!
